host/frontend/host-orchestrator/operator.go

```diff
@@ -0,0 +1,401 @@
+			if !ok {
+				log.Println("Device websocket closed")
+				return
+			}
+			if msg.Type != "forward" {
+				wsReplyError(ws, fmt.Sprintln("Unrecognized message type: ", msg.Type))
+				return
+			}
+			clientId := msg.ClientId
+			if clientId == 0 {
+				wsReplyError(ws, "Device forward message missing client id")
+				return
+			}
+			payload := msg.Payload
+			if payload == nil {
+				wsReplyError(ws, "Device forward message missing payload")
+				return
+			}
+			dMsg := map[string]interface{}{
+				"message_type": "device_msg",
+				"payload":      payload,
+			}
+			if !device.ToClient(clientId, dMsg) {
+				log.Println(ws, "Missing client ", int(clientId), " for device ", id)
+				wsReplyError(ws, fmt.Sprintln("Client disconnected: ", clientId))
+			}
+		}
+	}()
+}
+
+// General client endpoints
+
+func listDevices(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
+	if err := replyJSON(w, pool.DeviceIds()); err != nil {
+		log.Println(err)
+	}
+}
+
+func deviceFiles(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
+	vars := mux.Vars(r)
+	devId := vars["deviceId"]
+	dev := pool.GetDevice(devId)
+	if dev == nil {
+		http.NotFound(w, r)
+		return
+	}
+	path := vars["path"]
+	if shouldIntercept(path) {
+		http.ServeFile(w, r, fmt.Sprintf("intercept%s", path))
+	} else {
+		r.URL.Path = path
+		dev.Proxy.ServeHTTP(w, r)
+	}
+}
+
+// Client websocket endpoint
+
+func clientWs(w http.ResponseWriter, r *http.Request, pool *DevicePool, config InfraConfig) {
+	log.Println(r.URL)
+	ws := NewJsonWs(w, r)
+	if ws == nil {
+		return
+	}
+	// Serve the websocket in its own thread
+	go func() {
+		defer ws.Close()
+		var msg ConnectMsg
+		ok := ws.Recv(&msg)
+		if !ok {
+			log.Println("Websocket closed unexpectedly")
+			return
+		}
+		if msg.Type != "connect" {
+			wsReplyError(ws, "First client message must be 'connect'")
+			return
+		}
+		deviceId := msg.DeviceId
+		if deviceId == "" {
+			wsReplyError(ws, "Missing or invalid device_id")
+			return
+		}
+		device := pool.GetDevice(deviceId)
+		if device == nil {
+			wsReplyError(ws, fmt.Sprintln("Unknown device id: ", deviceId))
+			return
+		}
+		client := NewWsClient(ws)
+		id := device.Register(client)
+		defer device.Unregister(id)
+		if !client.Send(config) {
+			log.Println("Failed to send config to client")
+			return
+		}
+		infoMsg := make(map[string]interface{})
+		infoMsg["message_type"] = "device_info"
+		infoMsg["device_info"] = device.info
+		if !client.Send(infoMsg) {
+			log.Println("Failed to send device info to client")
+			return
+		}
+		for {
+			var msg ForwardMsg
+			ok := ws.Recv(&msg)
+			if !ok {
+				log.Println("Client websocket closed")
+				return
+			}
+			if msg.Type != "forward" {
+				wsReplyError(ws, fmt.Sprintln("Unrecognized message type: ", msg.Type))
+				return
+			}
+			payload := msg.Payload
+			if payload == nil {
+				wsReplyError(ws, "Client forward message missing payload")
+				return
+			}
+			cMsg := map[string]interface{}{
```

Comment:
Maybe worth creating a type for this message, considering it's used as well by `forward` below.

Reply:
Done